openmc/mgxs/mgxs.py

```diff
@@ -5891,3 +5897,466 @@ def get_units(self, xs_type='macro'):
+
+        # Construct a collection of the nuclides to retrieve from the xs tally
+        if self.by_nuclide:
+            if nuclides == 'all' or nuclides == 'sum' or nuclides == ['sum']:
+                query_nuclides = self.get_nuclides()
+            else:
+                query_nuclides = nuclides
+        else:
+            query_nuclides = ['total']
+
+        # If user requested the sum for all nuclides, use tally summation
+        if nuclides == 'sum' or nuclides == ['sum']:
+            xs_tally = self.xs_tally.summation(nuclides=query_nuclides)
+            xs = xs_tally.get_values(filters=filters,
+                                     filter_bins=filter_bins, value=value)
+        else:
+            xs = self.xs_tally.get_values(filters=filters,
+                                          filter_bins=filter_bins,
+                                          nuclides=query_nuclides, value=value)
+
+        # Divide by atom number densities for microscopic cross sections
+        if xs_type == 'micro' and self._divide_by_density:
+            if self.by_nuclide:
+                densities = self.get_nuclide_densities(nuclides)
+            else:
+                densities = self.get_nuclide_densities('sum')
+            if value == 'mean' or value == 'std_dev':
+                xs /= densities[np.newaxis, :, np.newaxis]
+
+        # Eliminate the trivial score dimension
+        xs = np.squeeze(xs, axis=len(xs.shape) - 1)
+        xs = np.nan_to_num(xs)
+
+        if groups == 'all':
+            num_groups = self.num_groups
+        else:
+            num_groups = len(groups)
+
+        # Reshape tally data array with separate axes for domain and energy
+        # Accomodate the polar and azimuthal bins if needed
+        num_surfaces = 4 * self.domain.n_dimension
+        num_subdomains = int(xs.shape[0] / (num_groups * self.num_polar *
+                                            self.num_azimuthal * num_surfaces))
+        if self.num_polar > 1 or self.num_azimuthal > 1:
+            new_shape = (self.num_polar, self.num_azimuthal, num_subdomains,
+                         num_groups, num_surfaces)
+        else:
+            new_shape = (num_subdomains, num_groups, num_surfaces)
+        new_shape += xs.shape[1:]
+        new_xs = np.zeros(new_shape)
+        for cell in range(num_subdomains):
+            for g in range(num_groups):
+                for s in range(num_surfaces):
+                    new_xs[cell,g,s] = \
+                            xs[cell*num_surfaces*num_groups+s*num_groups+g]
+        xs = new_xs
+
+        # Reverse data if user requested increasing energy groups since
+        # tally data is stored in order of increasing energies
+        if order_groups == 'increasing':
+            xs = xs[..., ::-1, :, :]
+
+        if squeeze:
+            # We want to squeeze out everything but the polar, azimuthal,
+            # and energy group data.
+            xs = self._squeeze_xs(xs)
+
+        return xs
```

Comment:
Instead of having this huge chunk of code repeating the get_xs in the MGXS class, just call the inherited one (with super) and reshape the returned xs to take into account the surfaces.

Reply:
I feel this could go either way. There are 3 other classes that redefine get_xs() to change the function in some ways, and not others. So to follow the same format, I would be inclined to keep this function since it is not identical. 
In this case 1/3 of the function is different, so I could justify leaving it as is.

Reply:
It could. If you remove the code for the per_nuclide and micro options then I'm fine with it, it wont be as much repeated code